### ISSUE TITLE
Integrate test dispatchers with libraries using Dispatchers.Xxx

### ISSUE THREAD

**Issue by elizarov** [Kotlin/kotlinx.coroutines#1365]
Make sure that libraries can use hard-coded `withContext(Dispatchers.IO) { ... }` and still be testable with test dispatchers. See also discussion in #810 and #1202.

**Comment by qwwdfsad** [comment]
The plan is to make `TestDefaultDispatcher` and `TestIoDispatcher` similar to `TestMainDispatcher` and integrate in into `kotlinx-coroutines-test` as soon as #1206 is here

**Comment by CharlyLafon37** [comment]
Any update on this issue ?

**Comment by shanling** [comment]
I too am curious if there's any update on this issue in 2020. Looking into solutions for testing `withContext(Dispatchers.IO/Default)` calls in my library that doesn't involve dependency injecting it into code where I don't actually want users to be able to provide their own dispatcher.

**Comment by matejdro** [comment]
Why don't you want users to provide your own dispatcher? Any solution here would basically be just another way to replace default dispatcher (just limited to the test library that could be used in production by your users).

**Comment by zach-klippenstein** [comment]
You can still use dependency injection without exposing to your consumers. E.g. an internal primary constructor that takes a Dispatcher, and a public secondary constructor that provides a default value. Or a public interface with internal implementation and a public factory function.

**Comment by CharlyLafon37** [comment]
@zach-klippenstein This solution does not help me with hardcoded ``withContext(Dispatchers.IO/Default)`` in libraries I have a dependency on

**Comment by zach-klippenstein** [comment]
Another potential approach to this problem, which still avoids overriding global singletons, would be to bake something like what is proposed [in this article](https://codingwithmohit.com/coroutines/custom-coroutine-context-uses-cases/) into the standard library, and change `Dispatchers.*` to instead simply indicate which of the `DispatcherProvider`-provided dispatchers to use (e.g. `withContext(Dispatchers.IO)` means lookup `coroutineContext[DispatcherProvider].io` and use that). Then any libraries that take a coroutine context automatically have configurable dispatchers. I'm sure there are still libraries that this still wouldn't solve for, but it feels less hacky for those that it would.

**Comment by matejdro** [comment]
This approach still requires you to have access to the scope/context. If library creates its scope internally, you cannot access that.

**Comment by BenTilbrook** [comment]
This issue keeps coming up for me. Is there anything we can do to help move it along?

**Comment by dkhalanskyjb** [comment]
This is (partially?) fixed now in 1.6.0-RC with the new API for testing.
* Using `withContext(Dispatchers.IO) { ... }` in `runTest` (which is a replacement for `runBlockingTest`) will wait for a minute for the code in the block to execute, so there won't be a near-instant crash.
* However, the delays are not skipped in `Dispatchers.IO`, `Dispatchers.Default`, and in general dispatchers that are not integrated with the test module. So, `withContext(Dispatchers.IO) { delay(100_000) }` will spin for a minute before crashing the test.

I'm not sure how much of a problem the remaining limitation is in practice: is it often that third-party code delays for long enough to notice? So, not closing this issue yet. Please share your thoughts and feedback!

**Comment by wkornewald** [comment]
We do have retry handling and other code that needs to work around limitations of underlying libraries with `delay` and those can happen inside non-Main dispatchers. It would be really great to have the possibility to not just set the main dispatcher but really all dispatchers. Currently we still depend on a wrapper (e.g. `dispatchers.io` instead of `Dispatchers.IO`) to have full control and a well-defined test environment. Does anything speak against just providing a simple `Dispatchers.setAll(dispatcher)`?

Also, it would be really awesome to have some utility to make setting up coroutine-capable unit tests trivially easy. Ideally, the default behavior in unit tests would automatically switch all dispatchers and the test lib would provide a global singleton default `testScope`. If anyone needs more control it's still possible to modify the default setup and use custom testScopes.

If having nice default behavior automatically is impossible, then it would be great to have something like a CoroutineTest base class or delegate or a test rule that can be added to any existing test class in as little code as possible to get the nice defaults. Anything that doesn't require repeating that boilerplate and learning unnecessary details just to get started would be great.

**Comment by sgc109** [comment]
> You can still use dependency injection without exposing to your consumers. E.g. an internal primary constructor that takes a Dispatcher, and a public secondary constructor that provides a default value. Or a public interface with internal implementation and a public factory function.

Could anyone explain more about what he is saying for me?

**Comment by dkhalanskyjb** [comment]
Not a very illustrating example. `spinTheWheel`, if it was part of the production code, is obviously buggy, so it's correct for the test to fail: for tests to be useful, they must reflect what's happening in production. That's what we're testing, the purpose of the process is not just to spin the CPU a bit and be satisfied with 15068/15068 tests being green.

So, our goal is twofold:
1. Make it difficult to write tests that pass when production fails,
2. Make it convenient to write tests that pass when production passes.

Making `Dispatchers.IO` deterministic in tests by default goes against the first goal, as your example shows, which is what holds us back. If we only recognized the second goal, then sure, you're right, not having to inject something is always more convenient.

Another issue with mocking `Dispatchers.IO` by default (or semi-default): there are some libraries that use `Dispatchers.IO` that *don't* expect to be mocked, (especially by a single-threaded dispatcher, which is another set of restrictions). For example, and this is a common occurrence, let's say you connect to some database in your tests. While most tests should probably mock the database calls, it's sometimes much more convenient (in cases of complex queries) and not any less robust to spawn an instance of some in-memory database just for running the tests. The following code becomes incorrect when run in an injected `Dispatchers.IO`: 
```kotlin
val connection = try {
  withTimeout(20.seconds) {
    database.connect()
  }
} catch (e: TimeoutCancellationException) {
  throw IllegalStateException("Couldn't connect to the database")
}
```
The reason is, those 20 seconds will pass instantly with the test framework, so the code will always time out. We would need a way for the author to say "I need to use `Dispatchers.IO` here, exactly the multithreaded version, the one without the delay-skipping, please don't mock it". Besides, in any case, is it a good idea to shift the burden on the author of this specific code so that they have to account for the delay skipping? We would be trading one group of disgruntled library authors who now have to make `Dispatchers.IO` explicitly injectable for another, those who would have to deal with the users of their library complaining that, with the mocking of `Dispatchers.IO` by the test framework, they see no way to have their code work well.

**Comment by francescotescari** [comment]
Thank you for your answer! What you say makes sense to me.
Let me just share that in my company (Android) in unit tests the dispatcher is always injected and mocked withh `TestDispatcher` (even `IO` and `Default`) and this seems to be aligned with what the guidelines of Android suggest ([1](https://developer.android.com/kotlin/coroutines/test), [2](https://developer.android.com/kotlin/coroutines/coroutines-best-practices)) and also seems to be a different approach compared to:

> Making Dispatchers.IO deterministic in tests by default goes against the first goal, as your example shows, which is what holds us back.

Regarding the database example issue, I see the problem and the problem it could be with third-party libraries.

**Comment by mash6699** [comment]
> This is (partially?) fixed now in 1.6.0-RC with the new API for testing.
> 
> * Using `withContext(Dispatchers.IO) { ... }` in `runTest` (which is a replacement for `runBlockingTest`) will wait for a minute for the code in the block to execute, so there won't be a near-instant crash. `withContext(Dispatchers.IO) { delay(100) }` will work just fine, for example.
> * However, the delays are not skipped in `Dispatchers.IO`, `Dispatchers.Default`, and in general dispatchers that are not integrated with the test module. So, `withContext(Dispatchers.IO) { delay(100_000) }` will spin for a minute before crashing the test.
> 
> I'm not sure how much of a problem the remaining limitation is in practice: is it often that third-party code delays for long enough to notice? So, not closing this issue yet. Please share your thoughts and feedback!

Thank you very much, this hack solve my issue

**Comment by trevjonez** [comment]
Can we bump the priority on this? Recent ART changes in Android 17 (API 37) have turned the original request into a forcing function, and it hasn't come up in the thread yet.

The standing answer on this issue has basically been "inject your dispatchers, or rely on `runTest` waiting on `withContext(Dispatchers.IO)`." That works when you own the code. The gap a lot of us have been living with is the code you *don't* own: a library that hardcodes `withContext(Dispatchers.IO)` or `Dispatchers.Default` internally and never exposes a seam to swap it. For those, the only thing that ever worked was reflectively replacing the field behind `Dispatchers.Default` / `Dispatchers.IO` / `Dispatchers.Unconfined`, the same trick people use to make Espresso idling resources see coroutine work. It was always a hack, and the docs on this issue rightly never blessed it, but in practice it was the one escape hatch.

On API 37 that escape hatch is gone. ART now enforces `static final` on writes at the native level, so stripping the `final` bit off the reflected field (the classic `accessFlags` patch) no longer has any effect and the write throws.

Testing on an API 37 emulator, here is what I have been able to discern:

| Target | Backing field | Result on API 37 |
| --- | --- | --- |
| `Dispatchers.Default` | `Dispatchers.Default` (static final) | `IllegalAccessException: Cannot set private static final field ...Dispatchers.Default` |
| `Dispatchers.Unconfined` | `Dispatchers.Unconfined` (static final) | same `IllegalAccessException` |
| `Dispatchers.IO` | `kotlinx.coroutines.scheduling.DefaultIoScheduler.default` (static final) | `IllegalAccessException: Cannot set private static final field ...DefaultIoScheduler.default` |
| any of the above via `MethodHandles.privateLookupIn(...).unreflectSetter` | | `IllegalAccessException: Field ... can't be overwritten` |

A couple of things worth calling out from that:

- After the `accessFlags` strip, `Field.getModifiers()` still reports `final == true`. The Java side mirror is cosmetic now; the check that matters is in ART and it isn't fooled.
- `Dispatchers.IO`'s backing field (`DefaultIoScheduler.default`) is itself `static final`, so it's in exactly the same boat as `Default`. If anyone has old test infra that assumed IO could be swapped through a non-final instance field, that assumption is dead on API 37 too.
- `Dispatchers.setMain()` still works, because it goes through the supported `TestMainDispatcher` machinery rather than reflection. So on API 37 you can still control Main and nothing else.

So the state of the world for library-internal dispatchers on API 37 is: no reflection and no `MethodHandles`. There is no practical way left to substitute a test dispatcher for a hardcoded `Dispatchers.IO`/`Default` in code you can't edit.

That's why I think API 37 turns this from a nice-to-have into a real hole. The "just inject it" guidance was always premised on the reflection hack quietly existing as the fallback for everyone else. Now the fallback is a hard `IllegalAccessException`, and the only remaining fix is the one @qwwdfsad described back in 2019 at the top of this thread: a supported `TestDefaultDispatcher` / `TestIoDispatcher` analogous to `TestMainDispatcher`, with `setDefault()` / `setIo()` entry points in `kotlinx-coroutines-test`.

**Comment by remcomokveld** [comment]
I think https://github.com/Kotlin/kotlinx.coroutines/issues/1365#issuecomment-628802412 is still a viable option as well. It is true that libraries would still be hard to test if they create their own scope, but if the library that could be fixed by initiating the library with something like `context: CoroutineContext = EmptyCoroutineContext` and the library uses that context as the base context for creating the scope.

That way you avoid the static mutability completely.

That would also further promote structured concurrency because you could pass the background scope context and the libraries scope would not leak past the lifetime of the test scope

**Comment by trevjonez** [comment]
That only cures half of my use case. The really big missing point for me isn't JVM unit tests but instrumentation tests running on android. I use the reflective hack to wrap the normal dispatcher in a counting idling resource so the test framework can globally track background work. That happens in the test harness before anything can really even use the Dispatchers class so by the time the test runs we have our hook in and registered. (anything that does happen to use it that early in the lifecycle will be inconsequential to the test runner behavior. IE: test harness internals the tests don't need to wait on, etc...)

Specifically it just delegates 100% of the API through to the normal dispatcher. Its only in my unit tests where I am doing the reflective hack to actually replace the behavior with the test dispatcher provided by runTest. Stuff running on the JVM as of at least 25 still works by letting mockito static mock hacks do the dirty work anyway. This is just a looming issue for ART at android 17 and up.

A funny anecdote to this reflective hackery is that it has actually prevented us from merging poorly behaving library code that was spamming work onto an IO dispatcher constantly (home grown OTEL library sampling way too much, tests would just hang and timeout waiting.) Ended up prompting a memory profiling session to also further block the same PR due to bad performance impacts on our apps. 

While I can concede that what we are doing sounds crazy at first glance, being able to swap things completely with the hack has served us incredibly well over the past 6+ years.

**Comment by JakeWharton** [comment]
Idling resources are a misguided strategy anyway since they don't actually capture the full cessation of work. You can't tell your mom on the phone "Wait until the `ExecutorService` is empty before clicking 'Submit'", it's just not how UI realistically settles.

Google has thankfully [changed their tune](https://medium.com/androiddevelopers/alternatives-to-idling-resources-in-compose-tests-8ae71f9fc473) on this, and the documentation [wants you to use `waitUntil` pattern](https://developer.android.com/training/testing/instrumented-tests/stability#prevent-synchronization) now.
